client.go

```diff
@@ -1355,7 +1367,34 @@ func (c *HostClient) acquireConn() (*clientConn, error) {
 		return cc, nil
 	}
 	if !createConn {
-		return nil, ErrNoFreeConns
+		if c.MaxConnWaitTimeout <= 0 {
+			return nil, ErrNoFreeConns
+		}
+
+		// wait for a free connection
+		ctx, cancelFunc := context.WithTimeout(context.Background(), c.MaxConnWaitTimeout)
+		w := &wantConn{
+			ready: make(chan struct{}, 1),
+		}
```

Comment:
Instead of using a context (which generates a lot of garbage) please use AcquireTimer. For example see:
https://github.com/valyala/fasthttp/blob/695f713fcf594b33fd8c126cfa8dc9f119d4391b/client.go#L782-L794

Reply:
Resolved, please see commit 014e836